CDAP-6156 Custom Namespace Mapping for Hive

rohitsinha54 opened the pull request:
Issue: https://issues.cask.co/browse/CDAP-6156
Build: http://builds.cask.co/browse/CDAP-DUT4348-1
ITN: http://builds.cask.co/browse/CDAP-ITM-7

Note: Opening it against branch feature/CDAP-6154-cust-ns-loc-map since the PR is based on the changes of that branch (PR: https://github.com/caskdata/cdap/pull/6009) to get the actual diff for review. This should be merged to develop. 
## Summary
- BaseHiveExplore uses NamespaceQueryAdmin to lookup namespace mapping for cdap namespaces. 
- Added a unit test which uses hive query "SHOW DATABASES" to check the custom database was created. 

TODO:
- [ ] Run ITN on the branch
- [ ] Test on cluster (Cluster creation is running. It has been tested on standalone)

albertshau committed:
CDAP-6375 refactor hydrator spark

Pulling Spark related Hydrator logic into its own module in
preparation for the new artifact that will run pipelines using
Spark Streaming.

As part of this refactor, a single Spark program can now be used
to run an entire pipeline, which removes the need for running the
planner logic and local connector datasets used by the mapreduce
engine.

PGuimarais committed:
Log table retrieves and populates from endpoint

Commit:
CDAP-6156 Custom Namespace Mapping for Hive

anwar6953 commented:
We need to avoid creation/deletion of hive db, if it's defined as external, right? 
I don't see that change in the other pr either

rohitsinha54 commented:
@anwar6953 : I have not done it since I don't know how to do it. I need some help from @poornachandra on that. I am adding comments from him to see.

rohitsinha54 commented:
@poornachandra : 
Create: 
https://github.com/caskdata/cdap/blob/develop/cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java#L742
In this create part if the database mapping was specified then I just want to check for the database existence. I didn't find any command which can check a database existence other than trying to do USE DATABASE and deducing from the result. Is there any other better way to check existence. 

Delete: 
https://github.com/caskdata/cdap/blob/develop/cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java#L771
I see that the delete returns a QueryHandle for custom mapping I will not run any query so can I just return null since I dont see anything being done with the return value 
https://github.com/caskdata/cdap/blob/develop/cdap-explore/src/main/java/co/cask/cdap/explore/executor/ExploreMetadataHttpHandler.java#L133

**But I think the simple and better way will be:**
Just have a simple if check and not call https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/StorageProviderNamespaceAdmin.java#L73
and
https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/StorageProviderNamespaceAdmin.java#L101
if a custom mapping was provided. The only issue with this is that it will work for delete but for create we do want to check the existence of the custom hive database. 

Thoughts ?

anwar6953 commented:
I think the check to avoid create/delete should be in the StorageProviderNamespaceAdmin.
And there, you would use the exploreFacade to check existence if its configured custom in the namespace config.

PGuimarais committed:
Adds timeline polling

elfenheart committed:
complete complex schema feature parity
update schema with implicit schema
add stream special case
add dataset selected event listener
remove clearDOM setting on schema clear event
fix error blink when switching to complex schema types
grouping schema types for simple and complex
handle stream import/export schema
cleanup commented out code
remove comment from dataset selected event
add spec-version 1.3
add error message when trying to export empty schema
modify compile to use es6 literals

PGuimarais committed:
Enables basic rendering of timeline data

PGuimarais committed:
Refactors link function

PGuimarais committed:
Refactors code and corrects timeline bug

PGuimarais committed:
Timeline scales with incoming logs

PGuimarais committed:
Restructures logviewer for single context

PGuimarais committed:
LogViewer and Timeline now follow the same time

poornachandra commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -1283,8 +1287,19 @@ private String getHiveDatabase(@Nullable String namespace) {
     }
-    String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
-    return namespace.equals(Id.Namespace.DEFAULT.getId()) ? namespace : String.format("%s_%s", tablePrefix, namespace);
+    try {
```
It would be good to have a comment describing what this piece of code does.

PGuimarais committed:
Resolves re-painting bug

PGuimarais committed:
resolves bug for killed services

poornachandra commented on cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java:
```diff
@@ -1283,8 +1287,19 @@ private String getHiveDatabase(@Nullable String namespace) {
+        return namespace;
+      } else if (!Strings.isNullOrEmpty(
+        namespaceQueryAdmin.get(Id.Namespace.from(namespace)).getConfig().getHiveDatabase())) {
+        return namespaceQueryAdmin.get(Id.Namespace.from(namespace)).getConfig().getHiveDatabase();
```
Second call not required, we can simply use the result of the first call, right?

PGuimarais committed:
Makes directive reusable

PGuimarais committed:
Removed commented code

PGuimarais committed:
Addresses comments

PGuimarais committed:
Addresses comments

PGuimarais committed:
Addresses comments

PGuimarais committed:
Remedies inconsistent rendering of tick lines

PGuimarais committed:
Fixes expand-all functionality

PGuimarais committed:
Adjusts table height to viewport and adds slider pin

PGuimarais committed:
Adjust positioning of pin

PGuimarais committed:
Adds ellipsis to truncated message text

PGuimarais committed:
Adds sliderHandle image

PGuimarais committed:
Adds basic infinite scrolling capabilities

PGuimarais committed:
Adds bootstrap well for loading data

PGuimarais committed:
Formats table cells

PGuimarais committed:
Adds tooltip to source column

astewes committed:
adds backend integration to Tags tab

PGuimarais committed:
Adds loading icon to bottom of table

PGuimarais committed:
Adds line number to tooltip

PGuimarais committed:
Infinite scroll now receives sequential data

PGuimarais committed:
rerenders on start time change

Commit:
[CDAP-6062] make PFS.dropPartition fail fast and abort if it cannot drop the partition from Hive; fix error propagation in explore service and client

astewes committed:
fixes styling issues with tags table + modals

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -130,13 +130,17 @@ public synchronized void create(NamespaceMeta metadata) throws Exception {
+    // TODO (CDAP-6155): this will be switched back to original when we do hbase mapping where we will start passing the
+    // NamespaceMeta itself to the DatasetFramework. We should first create namespaces in underlying storage before
+    // storing the namespace meta.
+    nsStore.create(metadata);
```
is this still being done? (the TODO)?

CuriousVini committed:
Adding support for multiple joiner plugin in pipeline

astewes committed:
fixes issues with pagination

astewes committed:
adds conditional truth meter info icon

gokulavasan committed:
Scaling RemoteSystemOperations by creating a new Guava service and start it in DatasetOpExecutorTwillRunnable

Should not be public-routable, used only through RemoteOpsClient

Add AuthModule to the DatasetOpTwillRunnable's injector since RemotePrivlege handler needs it

Adding RemoteSystemOp to system service status endpoint

Missed Inject annotation

Addressed comments

astewes committed:
adds minor layout fixes to results view

Commit:
Moved default implementations of enforce and filter to AbstractAuthorizer

Commit:
(CDAP-6127) set default bind port to 0 for tx service (#6078)

ajainarayanan committed:
Merge pull request #6019 from caskdata/feature/ui-new-log-viewer-integration

[CDAP-6174] Feature/ui new log viewer integration

sagarkapare committed:
CDAP-6528 Attempt Macro substituion only if the value is provided by user.

Commit:
CDAP-6154 Custom Namespace Mapping for Filesystem (#6009)

gokulavasan committed:
Merge pull request #6076 from caskdata/feature/remote-sysop-scale

Scaling RemoteSystemOperation handlers using DatasetOpExecutorTwillRunnable

albertshau committed:
Merge pull request #6065 from caskdata/feature/CDAP-6375-refactor-hydrator-spark

CDAP-6375 refactor hydrator spark

Commit:
Added AuthorizationEnforcementModule. Moved RemotePrivilegesFetcher to cdap-security and RemoteOpsClient/MethodArgument to cdap-common so they can be accessed by both cdap-app-fabric and cdap-data-fabric.

bdmogal committed:
Merge pull request #6081 from caskdata/feature/enforce-filter-abstract

Moved default implementations of enforce and filter to AbstractAuthor…

sagarkapare committed:
Merge pull request #6084 from caskdata/fix/itn-error

CDAP-6528 Attempt Macro substituion only if the value is provided by user.

bdmogal committed:
Merge pull request #6086 from caskdata/feature/move-classes

Added AuthorizationEnforcementModule. Moved RemotePrivilegesFetcher t…

astewes committed:
cleans up paths, removes unused empty

Commit:
[CDAP-6510] remove upgrade of explore - it was only for 2.6 -> 2.8

elfenheart committed:
Merge pull request #6022 from caskdata/feature/ui-complex-schema-parity

[CDAP-4986] UI Complex Schema Integration Part 2

astewes committed:
cleans up tag fetching + variable use in templates

astewes committed:
fixes vertical alignment of info icon

astewes committed:
Merge pull request #6071 from caskdata/feature/ui-tracker-add-meter-info-icon

adds conditional truth meter info icon

astewes committed:
Merge pull request #6072 from caskdata/ui-styling-tracker-fix-results-sort

Ui styling tracker fix results sort

Commit:
Merge branch 'feature/cdap-6510-remove-explore-upgrade' into develop

CuriousVini committed:
Merge pull request #6079 from caskdata/feature/multiphase-with-planner-fix

Adding support for multiple joiner plugin in pipeline

Commit:
Do not manage lifecycle of custom hive databases

rohitsinha54 commented:
Closing this. Please look at https://github.com/caskdata/cdap/pull/6096 which is against develop.
